### PR TITLE
refactor: migrate to alloy providers

### PR DESCRIPTION
**Pull request by onbjerg**
## Motivation

We want to migrate from Ethers to Alloy, which includes removing all references to Ether's `Middleware` trait and using Alloy equivalents (`alloy-providers`).

## Solution

The API for providers is not final, so a temporary trait has been introduced to make migrating to the final API easier later on. Additionally, I will be removing code that is no longer needed (e.g. `RuntimeClient`), as well as adjusting everything that breaks downstream as a result of me changing the providers (e.g. `forge script`).

## Todo

- [x] Adjust tests
- [x] Adjust docs
- [x] Resolve final breakages
- [x] Open up issues re: ergonomics with Alloy encountered during migration
- [x] Implement WebSocket and IPC support
- [x] Migrate Alchemy CU rate limiting mechanism as a Tower layer
- [x] Add retries as a tower layer on the transports

Depends on https://github.com/alloy-rs/alloy/pull/20

### CODE REVIEW

File: crates/common/src/provider/alloy.rs
Comment (onbjerg) [reply crates/common/src/provider/alloy.rs:207]: not relevant rn since we don't poll anything using this provider, but good to remember for later when we migrate cast

**Review by Evalir — APPROVED** [review]
Need to give this a final final review, but we should way until after this weekend to merge this to avoid any potential snafus with any small breakages we might face not covered by CI.

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/anvil/src/eth/backend/fork.rs
Comment (onbjerg) [reply crates/anvil/src/eth/backend/fork.rs:616]: doesn't exist in alloy and is not needed for anvil because we do not do subscriptions of any kind

**Comment by gakonst** [comment]
blocked on https://github.com/alloy-rs/alloy/pull/89 to ensure that we can move Reth Inspectors to use Alloy RPC instead of Reth RPC to avoid cyclical deps / incompats

**Comment by Evalir** [comment]
Unblocked thru https://github.com/alloy-rs/alloy/pull/89, https://github.com/alloy-rs/alloy/pull/96 & https://github.com/alloy-rs/alloy/pull/97, next step is to use the updated alloy-rpc-types to make this compile, and after, a rebase.

**Review by mattsse — CHANGES_REQUESTED** [review]
okay, we need another pass and remove alloy ethers conversion where possible

File: crates/anvil/src/eth/backend/mem/storage.rs
Comment (Evalir) [reply crates/anvil/src/eth/backend/mem/storage.rs:239]: These are storage-related types and we agreed not to migrate them on this PR.

File: crates/anvil/src/eth/api.rs
Comment (Evalir) [reply crates/anvil/src/eth/api.rs:2525]: Also an storage type.

File: crates/anvil/src/eth/api.rs
Comment (Evalir) [reply crates/anvil/src/eth/api.rs]: https://github.com/foundry-rs/foundry/pull/6219/commits/cba1d937b03ba464ea576a0c1f5dc6fc452027db

File: crates/anvil/src/pubsub.rs
Comment (Evalir) [reply crates/anvil/src/pubsub.rs]: https://github.com/foundry-rs/foundry/pull/6219/commits/2f3ebba0a9c5b0f6e8b87c54ef608617234494ae

File: crates/anvil/src/config.rs
Comment (Evalir) [reply crates/anvil/src/config.rs]: https://github.com/foundry-rs/foundry/pull/6219/commits/2f3ebba0a9c5b0f6e8b87c54ef608617234494ae

**Review by mattsse — APPROVED** [review]
more things to come!

**Comment by onbjerg** [comment]
I've tried looking at the ICE but I don't think there's a workaround. Should we merge with the ICE present?

**Comment by Evalir** [comment]
yep @onbjerg we can safely ignore the ICE for now—clippy won't run anyway until it's fixed.

**Pull request merged**
